Lookup introspection fails for global cached lookups

```
$ curl -v http://REDACTED_HOST/druid/v1/lookups/introspect/REDACTED/version
*   Trying REDACTED_HOST...
* Connected to REDACTED_HOST (REDACTED_HOST) port ##### (#0)
> GET /druid/v1/lookups/introspect/REDACTED_account_name_test/keys HTTP/1.1
> Host: REDACTED_HOST
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 500 Server Error
< Date: Mon, 27 Jun 2016 17:55:35 GMT
< Cache-Control: must-revalidate,no-cache,no-store
< Content-Type: text/html; charset=ISO-8859-1
< Content-Length: 614
< Connection: close
< Server: Jetty(9.2.z-SNAPSHOT)
<
<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 500 </title>
</head>
<body>
<h2>HTTP ERROR: 500</h2>
<p>Problem accessing /druid/v1/lookups/introspect/REDACTED/version. Reason:
<pre>    javax.servlet.ServletException: com.sun.jersey.api.container.ContainerException: java.lang.IllegalAccessException: Class com.sun.jersey.spi.container.JavaMethodInvokerFactory$1 can not access a member of class io.druid.query.lookup.NamespaceLookupExtractorFactory$1 with modifiers "public"</pre></p>
<hr /><i><small>Powered by Jetty://</small></i>
</body>
</html>
* Closing connection 0
```

This is a new feature for 0.9.1 so I don't think it should be a blocker. Suggesting labeling for 0.9.2

Full stacktrace:


```
javax.servlet.ServletException: com.sun.jersey.api.container.ContainerException: java.lang.IllegalAccessException: Class com.sun.jersey.spi.container.JavaMethodInvokerFactory$1 can not access a member of class io.druid.query.lookup.NamespaceLookupExtractorFactory$1 with modifiers "public"
        at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:423) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:558) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:733) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.google.inject.servlet.ServletDefinition.doServiceImpl(ServletDefinition.java:278) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.google.inject.servlet.ServletDefinition.doService(ServletDefinition.java:268) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.google.inject.servlet.ServletDefinition.service(ServletDefinition.java:180) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.google.inject.servlet.ManagedServletPipeline.service(ManagedServletPipeline.java:93) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.google.inject.servlet.ManagedFilterPipeline.dispatch(ManagedFilterPipeline.java:120) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.google.inject.servlet.GuiceFilter$1.call(GuiceFilter.java:132) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.google.inject.servlet.GuiceFilter$1.call(GuiceFilter.java:129) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.google.inject.servlet.GuiceFilter$Context.call(GuiceFilter.java:206) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:129) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.servlets.UserAgentFilter.doFilter(UserAgentFilter.java:83) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.servlets.GzipFilter.doFilter(GzipFilter.java:364) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1125) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1059) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.server.Server.handle(Server.java:497) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:248) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:620) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:540) [druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_45]
Caused by: com.sun.jersey.api.container.ContainerException: java.lang.IllegalAccessException: Class com.sun.jersey.spi.container.JavaMethodInvokerFactory$1 can not access a member of class io.druid.query.lookup.NamespaceLookupExtractorFactory$1 with modifiers "public"
        at com.sun.jersey.server.impl.model.method.dispatch.ResourceJavaMethodDispatcher.dispatch(ResourceJavaMethodDispatcher.java:84) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.HttpMethodRule.accept(HttpMethodRule.java:302) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.SubLocatorRule.accept(SubLocatorRule.java:137) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.ResourceClassRule.accept(ResourceClassRule.java:108) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.RootResourceClassesRule.accept(RootResourceClassesRule.java:84) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1542) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1473) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        ... 32 more
Caused by: java.lang.IllegalAccessException: Class com.sun.jersey.spi.container.JavaMethodInvokerFactory$1 can not access a member of class io.druid.query.lookup.NamespaceLookupExtractorFactory$1 with modifiers "public"
        at sun.reflect.Reflection.ensureMemberAccess(Reflection.java:102) ~[?:1.8.0_45]
        at java.lang.reflect.AccessibleObject.slowCheckMemberAccess(AccessibleObject.java:296) ~[?:1.8.0_45]
        at java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:288) ~[?:1.8.0_45]
        at java.lang.reflect.Method.invoke(Method.java:490) ~[?:1.8.0_45]
        at com.sun.jersey.spi.container.JavaMethodInvokerFactory$1.invoke(JavaMethodInvokerFactory.java:60) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.model.method.dispatch.AbstractResourceMethodDispatchProvider$ResponseOutInvoker._dispatch(AbstractResourceMethodDispatchProvider.java:205) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.model.method.dispatch.ResourceJavaMethodDispatcher.dispatch(ResourceJavaMethodDispatcher.java:75) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.HttpMethodRule.accept(HttpMethodRule.java:302) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.SubLocatorRule.accept(SubLocatorRule.java:137) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.ResourceClassRule.accept(ResourceClassRule.java:108) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.uri.rules.RootResourceClassesRule.accept(RootResourceClassesRule.java:84) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1542) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1473) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409) ~[druid-selfcontained-0.9.1-rc3-mmx0.jar:0.9.1-rc3-mmx0]
        ... 32 more
```

--- a/extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java
+++ b/extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java
@@ -47,8 +47,6 @@ import kafka.serializer.Decoder;
 
 import javax.annotation.Nullable;
 import javax.validation.constraints.Min;
-import javax.ws.rs.GET;
-import javax.ws.rs.core.Response;
 import java.nio.ByteBuffer;
 import java.util.List;
 import java.util.Map;
@@ -344,7 +342,7 @@ public class KafkaLookupExtractorFactory implements LookupExtractorFactory
   @Override
   public LookupIntrospectHandler getIntrospectHandler()
   {
-    return new KafkaLookupExtractorIntrospectionHandler();
+    return new KafkaLookupExtractorIntrospectionHandler(this);
   }
 
   @Override
@@ -405,20 +403,5 @@ public class KafkaLookupExtractorFactory implements LookupExtractorFactory
   ListenableFuture<?> getFuture()
   {
     return future;
-  }
-
-
-  class KafkaLookupExtractorIntrospectionHandler implements LookupIntrospectHandler
-  {
-    @GET
-    public Response getActive()
-    {
-      final ListenableFuture<?> future = getFuture();
-      if (future != null && !future.isDone()) {
-        return Response.ok().build();
-      } else {
-        return Response.status(Response.Status.GONE).build();
-      }
-    }
   }
 }

--- a/extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorIntrospectionHandler.java
+++ b/extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorIntrospectionHandler.java
@@ -1,0 +1,43 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.lookup;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.core.Response;
+
+public class KafkaLookupExtractorIntrospectionHandler implements LookupIntrospectHandler
+{
+  private KafkaLookupExtractorFactory kafkaLookupExtractorFactory;
+
+  public KafkaLookupExtractorIntrospectionHandler(KafkaLookupExtractorFactory kafkaLookupExtractorFactory) {this.kafkaLookupExtractorFactory = kafkaLookupExtractorFactory;}
+
+  @GET
+  public Response getActive()
+  {
+    final ListenableFuture<?> future = kafkaLookupExtractorFactory.getFuture();
+    if (future != null && !future.isDone()) {
+      return Response.ok().build();
+    } else {
+      return Response.status(Response.Status.GONE).build();
+    }
+  }
+}

--- a/extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/NamespaceLookupExtractorFactory.java
+++ b/extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/NamespaceLookupExtractorFactory.java
@@ -25,21 +25,14 @@ import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.annotation.JsonTypeName;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
-import com.google.common.collect.ImmutableMap;
 import com.metamx.common.ISE;
 import com.metamx.common.StringUtils;
 import com.metamx.common.logger.Logger;
-import io.druid.common.utils.ServletResourceUtils;
 import io.druid.query.extraction.MapLookupExtractor;
 import io.druid.query.lookup.namespace.ExtractionNamespace;
 import io.druid.server.lookup.namespace.cache.NamespaceExtractionCacheManager;
 
 import javax.annotation.Nullable;
-import javax.ws.rs.GET;
-import javax.ws.rs.Path;
-import javax.ws.rs.Produces;
-import javax.ws.rs.core.MediaType;
-import javax.ws.rs.core.Response;
 import java.nio.ByteBuffer;
 import java.util.Map;
 import java.util.UUID;
@@ -86,65 +79,7 @@ public class NamespaceLookupExtractorFactory implements LookupExtractorFactory
     this.injective = injective;
     this.manager = manager;
     this.extractorID = buildID();
-    this.lookupIntrospectHandler = new LookupIntrospectHandler()
-    {
-      @GET
-      @Path("/keys")
-      @Produces(MediaType.APPLICATION_JSON)
-      public Response getKeys()
-      {
-        try {
-          return Response.ok(getLatest().keySet()).build();
-        }
-        catch (ISE e) {
-          return Response.status(Response.Status.NOT_FOUND).entity(ServletResourceUtils.sanitizeException(e)).build();
-        }
-      }
-
-      @GET
-      @Path("/values")
-      @Produces(MediaType.APPLICATION_JSON)
-      public Response getValues()
-      {
-        try {
-          return Response.ok(getLatest().values()).build();
-        }
-        catch (ISE e) {
-          return Response.status(Response.Status.NOT_FOUND).entity(ServletResourceUtils.sanitizeException(e)).build();
-        }
-      }
-
-      @GET
-      @Path("/version")
-      @Produces(MediaType.APPLICATION_JSON)
-      public Response getVersion()
-      {
-        final String version = manager.getVersion(extractorID);
-        if (null == version) {
-          // Handle race between delete and this method being called
-          return Response.status(Response.Status.NOT_FOUND).build();
-        } else {
-          return Response.ok(ImmutableMap.of("version", version)).build();
-        }
-      }
-
-      @GET
-      @Produces(MediaType.APPLICATION_JSON)
-      public Response getMap()
-      {
-        try {
-          return Response.ok(getLatest()).build();
-        }
-        catch (ISE e) {
-          return Response.status(Response.Status.NOT_FOUND).entity(ServletResourceUtils.sanitizeException(e)).build();
-        }
-      }
-
-      private Map<String, String> getLatest()
-      {
-        return ((MapLookupExtractor) get()).getMap();
-      }
-    };
+    this.lookupIntrospectHandler = new NamespaceLookupIntrospectHandler(this, manager, extractorID);
   }
 
   @VisibleForTesting
@@ -166,13 +101,13 @@ public class NamespaceLookupExtractorFactory implements LookupExtractorFactory
         LOG.warn("Already started! [%s]", extractorID);
         return true;
       }
-      if(firstCacheTimeout > 0) {
+      if (firstCacheTimeout > 0) {
         if (!manager.scheduleAndWait(extractorID, extractionNamespace, firstCacheTimeout)) {
           LOG.error("Failed to schedule and wait for lookup [%s]", extractorID);
           return false;
         }
       } else {
-        if(!manager.scheduleOrUpdate(extractorID, extractionNamespace)) {
+        if (!manager.scheduleOrUpdate(extractorID, extractionNamespace)) {
           LOG.error("Failed to schedule lookup [%s]", extractorID);
           return false;
         }

--- a/extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/NamespaceLookupIntrospectHandler.java
+++ b/extensions-core/lookups-cached-global/src/main/java/io/druid/query/lookup/NamespaceLookupIntrospectHandler.java
@@ -1,0 +1,105 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.lookup;
+
+import com.google.common.collect.ImmutableMap;
+import com.metamx.common.ISE;
+import io.druid.common.utils.ServletResourceUtils;
+import io.druid.query.extraction.MapLookupExtractor;
+import io.druid.server.lookup.namespace.cache.NamespaceExtractionCacheManager;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.Map;
+
+public class NamespaceLookupIntrospectHandler implements LookupIntrospectHandler
+{
+  private final NamespaceLookupExtractorFactory factory;
+  private final String extractorID;
+  private final NamespaceExtractionCacheManager manager;
+  public NamespaceLookupIntrospectHandler(
+      NamespaceLookupExtractorFactory factory,
+      NamespaceExtractionCacheManager manager,
+      String extractorID
+  ) {
+    this.factory = factory;
+    this.extractorID = extractorID;
+    this.manager = manager;
+  }
+  @GET
+  @Path("/keys")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getKeys()
+  {
+    try {
+      return Response.ok(getLatest().keySet()).build();
+    }
+    catch (ISE e) {
+      return Response.status(Response.Status.NOT_FOUND).entity(ServletResourceUtils.sanitizeException(e)).build();
+    }
+  }
+
+  @GET
+  @Path("/values")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getValues()
+  {
+    try {
+      return Response.ok(getLatest().values()).build();
+    }
+    catch (ISE e) {
+      return Response.status(Response.Status.NOT_FOUND).entity(ServletResourceUtils.sanitizeException(e)).build();
+    }
+  }
+
+  @GET
+  @Path("/version")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getVersion()
+  {
+    final String version = manager.getVersion(extractorID);
+    if (null == version) {
+      // Handle race between delete and this method being called
+      return Response.status(Response.Status.NOT_FOUND).build();
+    } else {
+      return Response.ok(ImmutableMap.of("version", version)).build();
+    }
+  }
+
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getMap()
+  {
+    try {
+      return Response.ok(getLatest()).build();
+    }
+    catch (ISE e) {
+      return Response.status(Response.Status.NOT_FOUND).entity(ServletResourceUtils.sanitizeException(e)).build();
+    }
+  }
+
+  private Map<String, String> getLatest()
+  {
+    return ((MapLookupExtractor) factory.get()).getMap();
+  }
+}

